Core/automation/lib/javascript/core/log.js

```diff
@@ -0,0 +1,199 @@
+/**
+ * Logging functions and variables
+ * 
+ * Copyright (c) 2019 Contributors to the openHAB Scripters project
+ * 
+ * @author Martin Stangl - initial contribution
+ */
+'use strict';	
+
+(function context (context) {
+    'use strict';	
+    
+    load(__DIR__+'/actions.js');        
+
+    context.NOTIFY_OFF = 0;
+    context.NOTIFY_ERROR = 200;
+    context.NOTIFY_WARN = 300;
+    context.NOTIFY_INFO = 400;
+    context.NOTIFY_DEBUG = 500;
+
+    context.Logger = function Logger (name, notificationLevel, config) {        
+        if (name === undefined) {
+            name = Error().stack.split('\n')[2].split('/').slice(-2).join('.').split(':')[0];
+            name = name.slice(-3) == ".js" ? name.slice(0,-3) : null;
+        }
+        var _logger  = Java.type("org.slf4j.LoggerFactory").getLogger(name === null ? "jsr223.javascript" : "jsr223.javascript." + name.toString().toLowerCase());        
+
+        try {
+            // Set default config for config params not provided.
+            if (config === undefined) config = {};
+            if (config.ERROR === undefined) config.ERROR = {};
+            if (config.WARN  === undefined) config.WARN  = {};
+            if (config.INFO  === undefined) config.INFO  = {};
+            if (config.DEBUG === undefined) config.DEBUG = {};
+            if (config.ERROR.prefix === undefined) config.ERROR.prefix = "short";
+            if (config.WARN.prefix  === undefined) config.WARN.prefix  = "none";
+            if (config.INFO.prefix  === undefined) config.INFO.prefix  = "none";
+            if (config.DEBUG.prefix === undefined) config.DEBUG.prefix = "short";       
+
+            return Object.create(Object.prototype, {
+                _notificationLevel: { value: (notificationLevel === undefined || notificationLevel === null) ? context.NOTIFY_OFF : notificationLevel },
+                _config: { value: config },                
+                _name: { value: name },
+
+                error: { value: function error (msg) {
+                    try {
+                        if (_logger.isErrorEnabled()) {                      
+                            eval("_logger.error" + this._loggerArgs(this._getLogMessage(msg), arguments.length));                        
```

Comment:
Why do you construct the logging command as a string and then `eval` it, rather than formatting the message string and calling the function directly? Is there some magic going on, and if so should there be a comment explaining it?

Reply:
This eval is gone in the latest version. 
I could not figure out how to make it work otherwise initially. But the solution came to me later, when I was updating the notification part.